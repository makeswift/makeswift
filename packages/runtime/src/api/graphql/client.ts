export class GraphQLClient {
  private endpoint: string
  private headers: Record<string, string>

  constructor(endpoint: string, headers: Record<string, string> = {}) {
    this.endpoint = endpoint
    this.headers = headers
  }

  async request<TData, TVariables = Record<string, never>>(
    query: string,
    variables: TVariables = {} as TVariables,
  ): Promise<TData> {
    const response = await fetch(this.endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.headers,
      },
      body: JSON.stringify({ query, variables }),
    })

    if (!response.ok && response.status !== 400) {
      throw new Error(`${response.status} ${response.statusText}`)
    }

    const body = await response.json()

    if (body.errors != null) {
      console.error(body)

      throw new Error('GraphQL response contains errors, check the console.')
    }

    return body.data
  }
}
