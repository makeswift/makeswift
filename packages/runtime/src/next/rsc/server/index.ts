import 'server-only'

export { NextRSCMakeswiftPage as ExperimentalMakeswiftPage } from './makeswift-page'
export { NextRSCServerProvider as ExperimentalServerProvider } from './server-provider'
