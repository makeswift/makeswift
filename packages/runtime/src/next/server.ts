export { MakeswiftApiHandler } from './api-handler'
export type { Manifest, Font } from '../api-handler'
export type { MakeswiftApiHandlerResponse } from './api-handler'
export { getSiteVersion } from './draft-mode'
