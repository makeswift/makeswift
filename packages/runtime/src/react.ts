export type { Action } from './state/actions'
export { ActionTypes, changeDocument } from './state/actions'
export {
  Element,
  ReactRuntime,
  useIsInBuilder,
  DocumentRoot,
  PageProvider,
  usePageId,
} from './runtimes/react'
