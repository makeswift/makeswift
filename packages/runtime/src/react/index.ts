export { ReactRuntime } from '../runtimes/react'
