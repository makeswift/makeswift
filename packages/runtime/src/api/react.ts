import { type FetchableValue } from '@makeswift/controls'
import * as MakeswiftApiClient from '../state/makeswift-api-client'
import {
  APIResourceType,
  File,
  GlobalElement,
  LocalizedGlobalElement,
  Page,
  PagePathnameSlice,
  Site,
  Snippet,
  Swatch,
  Table,
  Typography,
} from './types'
import { GraphQLClient } from './graphql/client'
import { CreateTableRecordMutation } from './graphql/documents'
import {
  CreateTableRecordMutationResult,
  CreateTableRecordMutationVariables,
} from './graphql/generated/types'

export type SerializedLocalizedResourcesMap = {
  [resourceId: string]: string | null
}

type LocalizedResourcesMap = Map<string, string | null>

export type CacheData = MakeswiftApiClient.SerializedState

export type MakeswiftClientOptions = {
  uri: string
  cacheData?: CacheData
  localizedResourcesMap?: SerializedLocalizedResourcesMap
  locale?: Intl.Locale
}

/**
 * NOTE(miguel): This "client" is used to fetch Makeswift API resources needed for the host. For
 * example, swatches, files, typographies, etc. Ideally it's internal to the runtime and is only
 * used by controls to transform API references to API resources.
 *
 * Moreover, its use should be reserved for the builder only, since for live pages all Makeswift
 * API resources should be embedded in the "page snapshot". In the builder, this client serves the
 * purpose of sending requests for API resources and keeping a cache so that changes that happen in
 * the builder, like modifying a swatch, can be sent via `postMessage` to the host and the cache can
 * immediately update the value and re-render.
 *
 * Furthermore, the API resources requested shouldn't be requested directly from the Makeswift API
 * as that would require those resources to not be authenticated since the requests come from the
 * browser when running the host. Instead, the requests should go to the host directly, at the
 * Makeswift API endpoint (i.g., `/api/makeswift/[...makeswift]` dynamic route) where the host's
 * API key can be used, securely, in the server. For this reason, this client should really be a
 * client of the host's API, not Makeswift's, intended to build and continuously maintain a realtime
 * snapshot for use in the builder, not the lives pages.
 */
export class MakeswiftHostApiClient {
  graphqlClient: GraphQLClient
  makeswiftApiClient: MakeswiftApiClient.Store
  subscribe: MakeswiftApiClient.Store['subscribe']
  private localizedResourcesMap: LocalizedResourcesMap
  private locale: Intl.Locale | undefined

  constructor({ uri, cacheData, localizedResourcesMap = {}, locale }: MakeswiftClientOptions) {
    this.graphqlClient = new GraphQLClient(uri)
    this.makeswiftApiClient = MakeswiftApiClient.configureStore({ serializedState: cacheData })
    this.subscribe = this.makeswiftApiClient.subscribe
    this.localizedResourcesMap = new Map(Object.entries(localizedResourcesMap))
    this.locale = locale
  }

  readSwatch(swatchId: string): Swatch | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Swatch,
      swatchId,
    )
  }

  async fetchSwatch(swatchId: string): Promise<Swatch | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Swatch, swatchId),
    )
  }

  resolveSwatch(swatchId: string | undefined): FetchableValue<Swatch | null> {
    return this.resolveResource(APIResourceType.Swatch, {
      id: swatchId,
      read: id => this.readSwatch(id),
      fetch: id => this.fetchSwatch(id),
    })
  }

  readFile(fileId: string): File | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.File,
      fileId,
    )
  }

  async fetchFile(fileId: string): Promise<File | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.File, fileId),
    )
  }

  resolveFile(fileId: string | undefined): FetchableValue<File | null> {
    return this.resolveResource(APIResourceType.File, {
      id: fileId,
      read: id => this.readFile(id),
      fetch: id => this.fetchFile(id),
    })
  }

  readTypography(typographyId: string): Typography | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Typography,
      typographyId,
    )
  }

  async fetchTypography(typographyId: string): Promise<Typography | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Typography, typographyId),
    )
  }

  resolveTypography(typographyId: string | undefined): FetchableValue<Typography | null> {
    return this.resolveResource(APIResourceType.Typography, {
      id: typographyId,
      read: id => this.readTypography(id),
      fetch: id => this.fetchTypography(id),
    })
  }

  readGlobalElement(globalElementId: string): GlobalElement | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.GlobalElement,
      globalElementId,
    )
  }

  async fetchGlobalElement(globalElementId: string): Promise<GlobalElement | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.GlobalElement, globalElementId),
    )
  }

  readLocalizedGlobalElement(globalElementId: string): LocalizedGlobalElement | null {
    const localizedGlobalElementId = this.getLocalizedResourceId(globalElementId)

    if (localizedGlobalElementId == null) return null

    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.LocalizedGlobalElement,
      localizedGlobalElementId,
    )
  }

  async fetchLocalizedGlobalElement(
    globalElementId: string,
  ): Promise<LocalizedGlobalElement | null> {
    const locale = this.locale

    if (locale == null) return null

    // If getLocalizedResourceId returns null, it means we have tried to fetch the resource,
    // but the resource is not available. If we haven't fetched it yet, it'll return undefined.
    const noLocalizedResource = this.getLocalizedResourceId(globalElementId) === null

    if (noLocalizedResource) return null

    // TODO(fikri): We're checking the cache here because the cache hit will fail on fetchAPIResource.
    // This is because in the cache we're saving the ID of the localizedGlobalElementId,
    // but we're reading the cache using the globalElementId.
    // This is very confusing and it can lead to more bugs. We need to refactor how we handle
    // localized global element.
    const cacheResult = this.readLocalizedGlobalElement(globalElementId)

    if (cacheResult) return cacheResult

    const result = await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(
        APIResourceType.LocalizedGlobalElement,
        globalElementId,
        locale,
      ),
    )

    this.setLocalizedResourceId({
      resourceId: globalElementId,
      localizedResourceId: result?.id ?? null,
    })

    return result
  }

  readPagePathnameSlice(pageId: string): PagePathnameSlice | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.PagePathnameSlice,
      pageId,
    )
  }

  async fetchPagePathnameSlice(pageId: string): Promise<PagePathnameSlice | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.PagePathnameSlice, pageId, this.locale),
    )
  }

  resolvePagePathnameSlice(pageId: string | undefined): FetchableValue<PagePathnameSlice | null> {
    return this.resolveResource(APIResourceType.PagePathnameSlice, {
      id: pageId,
      read: id => this.readPagePathnameSlice(id),
      fetch: id => this.fetchPagePathnameSlice(id),
    })
  }

  resolveResource<R>(
    type: APIResourceType,
    {
      id,
      read,
      fetch,
    }: {
      id: string | undefined
      read: (id: string) => R | null
      fetch: (id: string) => Promise<R | null>
    },
  ): FetchableValue<R | null> {
    const _read = () => (id != null ? read(id) : null)
    let lastValue: R | null = null
    return {
      name: `${type}:${id}`,
      readStable: () => (lastValue = _read()),
      subscribe: (onUpdate: () => void) =>
        this.subscribe(() => {
          if (_read() !== lastValue) onUpdate()
        }),
      fetch: async () => (id != null ? fetch(id) : null),
    }
  }

  readTable(tableId: string): Table | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Table,
      tableId,
    )
  }

  async fetchTable(tableId: string): Promise<Table | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Table, tableId),
    )
  }

  async createTableRecord(tableId: string, columns: any): Promise<void> {
    await this.graphqlClient.request<
      CreateTableRecordMutationResult,
      CreateTableRecordMutationVariables
    >(CreateTableRecordMutation, { input: { data: { tableId, columns } } })
  }

  readSite(siteId: string): Site | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Site,
      siteId,
    )
  }

  readPage(pageId: string): Page | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Page,
      pageId,
    )
  }

  readSnippet(snippetId: string): Snippet | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Snippet,
      snippetId,
    )
  }

  private getLocalizedResourceId(resourceId: string): string | undefined | null {
    return this.localizedResourcesMap?.get(resourceId)
  }

  setLocalizedResourceId({
    resourceId,
    localizedResourceId,
  }: {
    resourceId: string
    localizedResourceId: string | null
  }): void {
    this.localizedResourcesMap.set(resourceId, localizedResourceId)
  }
}
