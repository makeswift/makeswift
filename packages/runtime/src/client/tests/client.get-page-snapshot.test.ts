import { MakeswiftClient, type MakeswiftPageDocument } from '../../client'
import { http, HttpResponse } from 'msw'

import { ReactRuntime } from '../../runtimes/react'

import { server } from '../../mocks/server'
import { TestOrigins, TestWorkingSiteVersion } from '../../testing/fixtures'

const TEST_API_KEY = 'myApiKey'
const { apiOrigin } = TestOrigins
const runtime = new ReactRuntime()

function createTestClient() {
  return new MakeswiftClient(TEST_API_KEY, { runtime, apiOrigin })
}

let consoleErrorSpy: jest.SpyInstance

const pageDocument: MakeswiftPageDocument = {
  id: '[page-id]',
  site: {
    id: '[site-id]',
  },
  data: {
    type: '[element-type]',
    key: '[element-key]',
    props: {},
  },
  snippets: [],
  fonts: [],
  meta: {},
  seo: {},
  localizedPages: [],
  locale: null,
}

beforeEach(() => {
  consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {})
})

afterEach(() => {
  consoleErrorSpy.mockRestore()
})

describe('getPageSnapshot', () => {
  const pathname = 'blog/hello-world'
  const snapshotUrl = `${apiOrigin}/v4/pages/${encodeURIComponent(pathname)}/document`
  const graphqlUrl = `${apiOrigin}/graphql`
  const locale = 'es-MX'

  test.each([undefined, 'fr', 'es-MX'])(
    'correctly passes `locale` parameter (%s)',
    async locale => {
      // Arrange
      const client = createTestClient()

      server.use(
        // page snapshot request
        http.get(
          snapshotUrl,
          ({ request }) => {
            const url = new URL(request.url)
            return HttpResponse.json(
              { ...pageDocument, locale: url.searchParams.get('locale') },
              { status: 200 },
            )
          },
          {
            once: true,
          },
        ),
        // introspection query
        http.post(graphqlUrl, () => HttpResponse.json({}, { status: 200 }), {
          once: true,
        }),
      )

      // Act
      const result = await client.getPageSnapshot(pathname, {
        locale,
        siteVersion: TestWorkingSiteVersion,
      })

      // Assert
      expect(result?.cacheData).toMatchSnapshot()
      expect(result?.document.locale).toEqual(locale ?? null)
    },
  )

  test.each([
    [undefined, '[fallback-page-id]'],
    [true, '[fallback-page-id]'],
    [false, '[page-id]'],
  ])(
    'correctly passes `allowLocaleFallback` parameter (%s)',
    async (allowLocaleFallback, expectedPageId) => {
      // Arrange
      const client = createTestClient()

      server.use(
        // page snapshot request
        http.get(
          snapshotUrl,
          ({ request }) => {
            const url = new URL(request.url)
            const allowFallback = url.searchParams.get('allowLocaleFallback') === 'true'
            return HttpResponse.json(
              {
                ...pageDocument,
                id: allowFallback ? '[fallback-page-id]' : '[page-id]',
              },
              { status: 200 },
            )
          },
          {
            once: true,
          },
        ),
        // introspection query
        http.post(graphqlUrl, () => HttpResponse.json({}, { status: 200 }), {
          once: true,
        }),
      )

      // Act
      const result = await client.getPageSnapshot(pathname, {
        locale,
        allowLocaleFallback,
        siteVersion: TestWorkingSiteVersion,
      })

      // Assert
      expect(result?.document.id).toEqual(expectedPageId)
    },
  )

  test('returns null on 404', async () => {
    // Arrange
    const client = createTestClient()

    server.use(
      http.get(snapshotUrl, () => HttpResponse.text('', { status: 404 }), {
        once: true,
      }),
    )

    // Act
    const result = await client.getPageSnapshot(pathname, {
      locale,
      siteVersion: TestWorkingSiteVersion,
    })

    // Assert
    expect(result).toBeNull()
    expect(consoleErrorSpy).not.toHaveBeenCalled()
  })

  test('throws on errors other than 404, logs details to the console', async () => {
    // Arrange
    const client = createTestClient()

    server.use(
      http.get(snapshotUrl, () => HttpResponse.text('Internal server error', { status: 500 }), {
        once: true,
      }),
    )

    // Act
    const resultPromise = client.getPageSnapshot(pathname, {
      siteVersion: null,
      locale,
    })

    try {
      await resultPromise
    } catch (e) {}

    // Assert
    expect(resultPromise).rejects.toThrow(
      "Failed to get page snapshot for 'blog/hello-world': 500 Internal Server Error",
    )

    expect(consoleErrorSpy).toHaveBeenCalledWith(
      "Failed to get page snapshot for 'blog/hello-world'",
      {
        response: 'Internal server error',
        siteVersion: null,
        locale: 'es-MX',
      },
    )
  })
})
