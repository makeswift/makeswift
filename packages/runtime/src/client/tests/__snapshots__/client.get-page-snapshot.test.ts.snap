// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPageSnapshot correctly passes \`locale\` parameter (es-MX) 1`] = `
{
  "apiResources": {
    "File": [],
    "GlobalElement": [],
    "LocalizedGlobalElement": [],
    "PagePathnameSlice": [],
    "Swatch": [],
    "Table": [],
    "Typography": [],
  },
  "localizedResourcesMap": {
    "es-MX": {},
  },
}
`;

exports[`getPageSnapshot correctly passes \`locale\` parameter (fr) 1`] = `
{
  "apiResources": {
    "File": [],
    "GlobalElement": [],
    "LocalizedGlobalElement": [],
    "PagePathnameSlice": [],
    "Swatch": [],
    "Table": [],
    "Typography": [],
  },
  "localizedResourcesMap": {
    "fr": {},
  },
}
`;

exports[`getPageSnapshot correctly passes \`locale\` parameter (undefined) 1`] = `
{
  "apiResources": {
    "File": [],
    "GlobalElement": [],
    "LocalizedGlobalElement": [],
    "PagePathnameSlice": [],
    "Swatch": [],
    "Table": [],
    "Typography": [],
  },
  "localizedResourcesMap": {},
}
`;
